allennlp/modules/bimpm_matching.py

```diff
@@ -0,0 +1,361 @@
+    assert weight.size(1) == vector1.size(2) == vector1.size(2)
+
+    # (batch, seq_len, 1)
+    similarity_single = F.cosine_similarity(vector1, vector2, 2).unsqueeze(2)
+
+    # (1, 1, num_perspective, hidden_size)
+    weight = weight.unsqueeze(0).unsqueeze(0)
+
+    # (batch, seq_len, num_perspective, hidden_size)
+    vector1 = weight * vector1.unsqueeze(2)
+    vector2 = weight * vector2.unsqueeze(2)
+
+    similarity_multi = F.cosine_similarity(vector1, vector2, dim=3)
+
+    return similarity_single, similarity_multi
+
+
+def multi_perspective_match_pairwise(vector1: torch.Tensor,
+                                     vector2: torch.Tensor,
+                                     weight: torch.Tensor,
+                                     eps: float = 1e-8) -> torch.Tensor:
+    """
+    Calculate multi-perspective cosine matching between each time step of
+    one vector and each time step of another vector.
+
+    Parameters
+    ----------
+    vector1 : ``torch.Tensor``
+        A tensor of shape ``(batch, seq_len1, hidden_size)``
+    vector2 : ``torch.Tensor``
+        A tensor of shape ``(batch, seq_len2, hidden_size)``
+    weight : ``torch.Tensor``
+        A tensor of shape ``(num_perspective, hidden_size)``
+    eps : ``float`` optional, (default = 1e-8)
+        A small value to avoid zero division problem
+
+    Returns
+    -------
+    A tensor of shape (batch, seq_len1, seq_len2, num_perspective) consisting
+    multi-perspective matching results
+    """
+    num_perspective = weight.size(0)
+
+    # (1, num_perspective, 1, hidden_size)
+    weight = weight.unsqueeze(0).unsqueeze(2)
+
+    # (batch, num_perspective, seq_len*, hidden_size)
+    vector1 = weight * vector1.unsqueeze(1).expand(-1, num_perspective, -1, -1)
+    vector2 = weight * vector2.unsqueeze(1).expand(-1, num_perspective, -1, -1)
+
+    # (batch, num_perspective, seq_len*, 1)
+    vector1_norm = vector1.norm(p=2, dim=3, keepdim=True)
+    vector2_norm = vector2.norm(p=2, dim=3, keepdim=True)
+
+    # (batch, num_perspective, seq_len1, seq_len2)
+    mul_result = torch.matmul(vector1, vector2.transpose(2, 3))
+    norm_value = vector1_norm * vector2_norm.transpose(2, 3)
+
+    # (batch, seq_len1, seq_len2, num_perspective)
+    return (mul_result / norm_value.clamp(min=eps)).permute(0, 2, 3, 1)
+
+
+class BiMpmMatching(nn.Module, FromParams):
+    """
+    This ``Module`` implements the matching layer of BiMPM model described in `Bilateral
+    Multi-Perspective Matching for Natural Language Sentences <https://arxiv.org/abs/1702.03814>`_
+    by Zhiguo Wang et al., 2017.
+    Also please refer to the `TensorFlow implementation <https://github.com/zhiguowang/BiMPM/>`_ and
+    `PyTorch implementation <https://github.com/galsang/BIMPM-pytorch>`_.
+
+    Parameters
+    ----------
+    hidden_dim : ``int``, optional (default = 100)
+        The hidden dimension of the representations
+    num_perspective : ``int``, optional (default = 20)
```

Comment:
Nit: this should probably be `num_perspectives`, here, in the description on the next line, and as a variable name in the code..

Reply:
Sure, I will rename it.